File: examples/dekita/.claude/hooks/systematization_issue_close_check.py
#!/usr/bin/env python3
"""仕組み化Issueクローズ時にフック/ツール実装を検証。

Why:
    「仕組み化」を謳うIssueがドキュメント追加のみでクローズされると、
    強制機構がなく問題が再発する。クローズ前に実装を確認する。

What:
    - gh issue close時（PreToolUse:Bash）に発火
    - Issue内容から「仕組み化」系キーワードを検出
    - 関連PRで強制機構ファイル（hooks/workflows/scripts）の変更を確認
    - 強制機構がない場合はクローズをブロック

Remarks:
    - ブロック型フック（強制機構なしの場合はブロック）
    - systematization-checkはセッション終了時、本フックはIssueクローズ時
    - SKIP_SYSTEMATIZATION_CHECK=1でスキップ可能

Changelog:
    - silenvx/dekita#1909: フック追加
    - silenvx/dekita#2607: HookContextパターン移行
"""

import json
import os
import re
import subprocess
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent))

from lib.constants import TIMEOUT_MEDIUM
from lib.execution import log_hook_execution
from lib.results import make_block_result, print_continue_and_log_skip
from lib.session import create_hook_context, parse_hook_input
from lib.strings import extract_inline_skip_env, is_skip_env_enabled, strip_quoted_strings

HOOK_NAME = "systematization-issue-close-check"

# 「仕組み化」系キーワード
SYSTEMATIZATION_KEYWORDS = [
    r"仕組み化",
    r"フック(?:を)?(?:作成|追加|実装)",
    r"hook(?:を)?(?:作成|追加|実装)",
    r"CI(?:を)?(?:追加|実装)",
    r"自動(?:化|チェック)",
    r"強制(?:機構|チェック)",
    r"再発防止(?:策)?(?:の)?(?:仕組み|フック)",
]

# 強制機構ファイルのパターン
ENFORCEMENT_FILE_PATTERNS = [
    r"\.claude/hooks/.*\.py$",
    r"\.github/workflows/.*\.ya?ml$",
    r"\.claude/scripts/.*\.(?:py|sh)$",
]


def extract_issue_number(command: str) -> str | None:
    """gh issue close コマンドからIssue番号を抽出."""
    cmd = strip_quoted_strings(command)

    if not re.search(r"gh\s+issue\s+close\b", cmd):
        return None

    match = re.search(r"gh\s+issue\s+close\s+(.+)", cmd)
    if not match:
        return None

    args = match.group(1)

    for part in args.split():
        if part.startswith("-"):
            continue
        num_match = re.match(r"#?(\d+)$", part)
        if num_match:
            return num_match.group(1)

    return None


def get_issue_content(issue_number: str) -> tuple[str, str] | None:
    """IssueのタイトルとボディをGitHubから取得.

    Returns:
        (title, body) のタプル、取得失敗時は None
    """
    try:
        result = subprocess.run(
            [
                "gh",
                "issue",
                "view",
                issue_number,
                "--json",
                "title,body",
            ],
            capture_output=True,
            text=True,
            timeout=TIMEOUT_MEDIUM,
            check=False,
        )

        if result.returncode != 0:
            return None

        data = json.loads(result.stdout)
        return data.get("title", ""), data.get("body", "")

    except (subprocess.TimeoutExpired, OSError, json.JSONDecodeError):
        return None


def has_systematization_keyword(title: str, body: str) -> bool:
    """タイトルまたはボディに「仕組み化」系キーワードがあるか."""
    text = f"{title}\n{body}"
    for pattern in SYSTEMATIZATION_KEYWORDS:
        if re.search(pattern, text, re.IGNORECASE):
            return True
    return False


def get_pr_files(pr_number: int | str) -> list[str]:
    """PRの変更ファイル一覧を取得."""
    files: list[str] = []
    try:
        files_result = subprocess.run(
            [
                "gh",
                "pr",
                "view",
                str(pr_number),
                "--json",
                "files",
                "--jq",
                ".files[].path",
            ],
            capture_output=True,
            text=True,
            timeout=TIMEOUT_MEDIUM,
            check=False,
        )

        if files_result.returncode == 0:
            for line in files_result.stdout.strip().split("\n"):
                if line:
                    files.append(line)
    except (subprocess.TimeoutExpired, OSError):
        # gh CLI失敗時は空リストを返す（ベストエフォート）
        pass

    return files


def search_prs_by_issue(issue_number: str) -> list[int]:
    """Issue番号でPRを検索（open/merged両方）.

    Returns:
        見つかったPR番号のリスト（最新順）
    """
    pr_numbers: list[int] = []

    try:
        # --state all でopen/closed/merged全てを検索
        result = subprocess.run(
            [
                "gh",
                "pr",
                "list",
                "--state",
                "all",
                "--search",
                issue_number,
                "--json",
                "number",
                "--jq",
                ".[] | .number",
            ],
            capture_output=True,
            text=True,
            timeout=TIMEOUT_MEDIUM,
            check=False,
        )

        if result.returncode == 0:
            for line in result.stdout.strip().split("\n"):
                if line and line.isdigit():
                    pr_numbers.append(int(line))

    except (subprocess.TimeoutExpired, OSError):
        # gh CLI失敗時は空リストを返す（ベストエフォート）
        pass

    return pr_numbers


def get_linked_pr_files(issue_number: str) -> list[str]:
    """Issueに紐づくPRで変更されたファイル一覧を取得.

    1. まずlinkedPullRequestsを確認
    2. 見つからない場合はIssue番号でPR検索（マージ済み含む）
    """
    files: list[str] = []
    pr_numbers: list[int] = []

    try:
        # 1. Issueに正式にリンクされたPRの情報を取得
        result = subprocess.run(
            [
                "gh",
                "issue",
                "view",
                issue_number,
                "--json",
                "linkedPullRequests",
            ],
            capture_output=True,
            text=True,
            timeout=TIMEOUT_MEDIUM,
            check=False,
        )

        if result.returncode == 0:
            issue_data = json.loads(result.stdout)
            prs = issue_data.get("linkedPullRequests", [])
            for pr in prs:
                pr_num = pr.get("number")
                if pr_num:
                    pr_numbers.append(pr_num)

    except (subprocess.TimeoutExpired, OSError, json.JSONDecodeError):
        # gh CLI失敗時は空リストで続行（フォールバック検索へ）
        pass

    # 2. linkedPullRequestsが空の場合、Issue番号でPR検索
    if not pr_numbers:
        pr_numbers = search_prs_by_issue(issue_number)

    # 3. 各PRの変更ファイルを取得
    for pr_number in pr_numbers:
        pr_files = get_pr_files(pr_number)
        files.extend(pr_files)

    return files


def has_enforcement_file(files: list[str]) -> list[str]:
    """変更ファイルリストから強制機構ファイルを検出.

    Returns:
        強制機構ファイルのリスト
    """
    enforcement_files = []
    for file_path in files:
        for pattern in ENFORCEMENT_FILE_PATTERNS:
            if re.search(pattern, file_path):
                enforcement_files.append(file_path)
                break
    return enforcement_files


def main() -> None:
    """フックのエントリポイント."""
    result = {"decision": "approve"}

    try:
        input_data = parse_hook_input()
        # Issue #2607: Create context for session_id logging
        ctx = create_hook_context(input_data)
        tool_name = input_data.get("tool_name", "")

        if tool_name != "Bash":
            print_continue_and_log_skip(HOOK_NAME, f"not Bash: {tool_name}", ctx=ctx)
            return

        tool_input = input_data.get("tool_input", {})
        command = tool_input.get("command", "")

        # スキップ環境変数のチェック
        skip_env_name = "SKIP_SYSTEMATIZATION_CHECK"
        if is_skip_env_enabled(os.environ.get(skip_env_name)):
            log_hook_execution(
                HOOK_NAME,
                "approve",
                f"{skip_env_name} でスキップ（環境変数）",
            )
            print(json.dumps(result))
            return

        inline_value = extract_inline_skip_env(command, skip_env_name)
        if is_skip_env_enabled(inline_value):
            log_hook_execution(
                HOOK_NAME,
                "approve",
                f"{skip_env_name} でスキップ（インライン）",
            )
            print(json.dumps(result))
            return

        # gh issue close コマンドを検出
        issue_number = extract_issue_number(command)
        if not issue_number:
            print_continue_and_log_skip(HOOK_NAME, "no issue number found", ctx=ctx)
            return

        # Issue内容を取得
        content = get_issue_content(issue_number)
        if not content:
            log_hook_execution(
                HOOK_NAME,
                "approve",
                f"Issue #{issue_number} の内容取得失敗",
            )
            print(json.dumps(result))
            return

        title, body = content

        # ドキュメント更新Issueは対象外（docs: または docs(...) プレフィックス）
        if re.match(r"^docs[:\(]", title, re.IGNORECASE):
            log_hook_execution(
                HOOK_NAME,
                "approve",
                f"Issue #{issue_number} はドキュメント更新Issue（対象外）",
            )
            print(json.dumps(result))
            return

        # 「仕組み化」系キーワードをチェック
        if not has_systematization_keyword(title, body):
            log_hook_execution(
                HOOK_NAME,
                "approve",
                f"Issue #{issue_number} は仕組み化Issueではない",
            )
            print(json.dumps(result))
            return

        # 関連PRの変更ファイルを取得
        pr_files = get_linked_pr_files(issue_number)

        # 強制機構ファイルがあるかチェック
        enforcement_files = has_enforcement_file(pr_files)

        if enforcement_files:
            files_list = ", ".join(enforcement_files[:3])
            if len(enforcement_files) > 3:
                files_list += f" 他{len(enforcement_files) - 3}件"
            log_hook_execution(
                HOOK_NAME,
                "approve",
                f"Issue #{issue_number} に強制機構ファイルあり: {files_list}",
            )
            print(json.dumps(result))
            return

        # 強制機構ファイルがない場合はブロック
        reason_lines = [
            f"Issue #{issue_number} は「仕組み化」を謳っていますが、",
            "強制機構（フック/CI/ツール）の実装が確認できません。",
            "",
            "**仕組み化の定義**:",
            "  - ドキュメント追加だけでは不十分",
            "  - 違反を**ブロック**するフック/CI/ツールが必要",
            "",
            "**確認された変更ファイル**:",
        ]

        if pr_files:
            for f in pr_files[:5]:
                reason_lines.append(f"  - {f}")
            if len(pr_files) > 5:
                reason_lines.append(f"  ... 他 {len(pr_files) - 5} 件")
        else:
            reason_lines.append("  (関連PRが見つからないか、変更ファイルがありません)")

        reason_lines.extend(
            [
                "",
                "**対応方法**:",
                "  1. フック/CI/スクリプトを実装してからクローズ",
                "  2. 強制機構が不要と判断した場合はコメントで理由を説明",
                "",
                "**スキップ方法（確認済みの場合）**:",
                f"  SKIP_SYSTEMATIZATION_CHECK=1 gh issue close {issue_number}",
            ]
        )

        reason = "\n".join(reason_lines)
        result = make_block_result(HOOK_NAME, reason)

        log_hook_execution(
            HOOK_NAME,
            "block",
            f"Issue #{issue_number} に強制機構ファイルなし",
        )

    except Exception as e:
        result = {"decision": "approve"}
        log_hook_execution(HOOK_NAME, "error", f"フックエラー: {e}")

    print(json.dumps(result))


if __name__ == "__main__":
    main()

File: examples/dekita/.claude/hooks/handlers/systematization_issue_close_check.ts
#!/usr/bin/env bun
/**
 * 仕組み化Issueクローズ時にフック/ツール実装を検証。
 *
 * Why:
 *   「仕組み化」を謳うIssueがドキュメント追加のみでクローズされると、
 *   強制機構がなく問題が再発する。クローズ前に実装を確認する。
 *
 * What:
 *   - gh issue close時（PreToolUse:Bash）に発火
 *   - Issue内容から「仕組み化」系キーワードを検出
 *   - 関連PRで強制機構ファイル（hooks/workflows/scripts）の変更を確認
 *   - 強制機構がない場合はクローズをブロック
 *
 * Remarks:
 *   - ブロック型フック（強制機構なしの場合はブロック）
 *   - systematization-checkはセッション終了時、本フックはIssueクローズ時
 *   - SKIP_SYSTEMATIZATION_CHECK=1でスキップ可能
 *   - Python版: systematization_issue_close_check.py
 *
 * Changelog:
 *   - silenvx/dekita#1909: フック追加
 *   - silenvx/dekita#2607: HookContextパターン移行
 *   - silenvx/dekita#2917: TypeScript版初期実装
 */

import { execSync } from "node:child_process";
import { TIMEOUT_MEDIUM } from "../lib/constants";
import { formatError } from "../lib/format_error";
import { logHookExecution } from "../lib/logging";
import { makeApproveResult, makeBlockResult } from "../lib/results";
import { parseHookInput } from "../lib/session";
import { extractInlineSkipEnv, isSkipEnvEnabled, stripQuotedStrings } from "../lib/strings";

const HOOK_NAME = "systematization-issue-close-check";
const SKIP_ENV_NAME = "SKIP_SYSTEMATIZATION_CHECK";

// 「仕組み化」系キーワード
const SYSTEMATIZATION_KEYWORDS = [
  /仕組み化/i,
  /フック(?:を)?(?:作成|追加|実装)/i,
  /hook(?:を)?(?:作成|追加|実装)/i,
  /CI(?:を)?(?:追加|実装)/i,
  /自動(?:化|チェック)/i,
  /強制(?:機構|チェック)/i,
  /再発防止(?:策)?(?:の)?(?:仕組み|フック)/i,
];

// 強制機構ファイルのパターン
const ENFORCEMENT_FILE_PATTERNS = [
  /\.claude\/hooks\/.*\.py$/,
  /\.claude\/hooks\/.*\.ts$/,
  /\.github\/workflows\/.*\.ya?ml$/,
  /\.claude\/scripts\/.*\.(?:py|sh|ts)$/,
];

/**
 * Extract issue number from gh issue close command.
 */
export function extractIssueNumber(command: string): string | null {
  const cmd = stripQuotedStrings(command);

  if (!/gh\s+issue\s+close\b/.test(cmd)) {
    return null;
  }

  const match = cmd.match(/gh\s+issue\s+close\s+(.+)/);
  if (!match) {
    return null;
  }

  const args = match[1];

  for (const part of args.split(/\s+/)) {
    if (part.startsWith("-")) {
      continue;
    }
    const numMatch = part.match(/^#?(\d+)$/);
    if (numMatch) {
      return numMatch[1];
    }
  }

  return null;
}

/**
 * Get issue title and body from GitHub.
 */
function getIssueContent(issueNumber: string): { title: string; body: string } | null {
  try {
    const result = execSync(`gh issue view ${issueNumber} --json title,body`, {
      encoding: "utf-8",
      timeout: TIMEOUT_MEDIUM * 1000,
      stdio: ["pipe", "pipe", "pipe"],
    });

    const data = JSON.parse(result);
    return {
      title: data.title || "",
      body: data.body || "",
    };
  } catch {
    return null;
  }
}

/**
 * Check if title or body contains systematization keywords.
 */
export function hasSystematizationKeyword(title: string, body: string): boolean {
  const text = `${title}\n${body}`;
  for (const pattern of SYSTEMATIZATION_KEYWORDS) {
    if (pattern.test(text)) {
      return true;
    }
  }
  return false;
}

/**
 * Get list of files changed in a PR.
 */
function getPrFiles(prNumber: number | string): string[] {
  const files: string[] = [];
  try {
    const result = execSync(`gh pr view ${prNumber} --json files --jq '.files[].path'`, {
      encoding: "utf-8",
      timeout: TIMEOUT_MEDIUM * 1000,
      stdio: ["pipe", "pipe", "pipe"],
    });

    for (const line of result.trim().split("\n")) {
      if (line) {
        files.push(line);
      }
    }
  } catch {
    // gh CLI failure - return empty list (best-effort)
  }
  return files;
}

/**
 * Search PRs by issue number (open/merged both).
 */
function searchPrsByIssue(issueNumber: string): number[] {
  const prNumbers: number[] = [];
  try {
    const result = execSync(
      `gh pr list --state all --search ${issueNumber} --json number --jq '.[] | .number'`,
      {
        encoding: "utf-8",
        timeout: TIMEOUT_MEDIUM * 1000,
        stdio: ["pipe", "pipe", "pipe"],
      },
    );

    for (const line of result.trim().split("\n")) {
      if (line && /^\d+$/.test(line)) {
        prNumbers.push(Number.parseInt(line, 10));
      }
    }
  } catch {
    // gh CLI failure - return empty list (best-effort)
  }
  return prNumbers;
}

/**
 * Get files changed in PRs linked to the issue.
 */
function getLinkedPrFiles(issueNumber: string): string[] {
  const files: string[] = [];
  let prNumbers: number[] = [];

  try {
    // 1. Get officially linked PRs
    const result = execSync(`gh issue view ${issueNumber} --json linkedPullRequests`, {
      encoding: "utf-8",
      timeout: TIMEOUT_MEDIUM * 1000,
      stdio: ["pipe", "pipe", "pipe"],
    });

    const data = JSON.parse(result);
    const prs = data.linkedPullRequests || [];
    for (const pr of prs) {
      if (pr.number) {
        prNumbers.push(pr.number);
      }
    }
  } catch {
    // Continue to fallback search
  }

  // 2. If no linked PRs found, search by issue number
  if (prNumbers.length === 0) {
    prNumbers = searchPrsByIssue(issueNumber);
  }

  // 3. Get files from each PR
  for (const prNumber of prNumbers) {
    const prFiles = getPrFiles(prNumber);
    files.push(...prFiles);
  }

  return files;
}

/**
 * Find enforcement files from the list of changed files.
 */
export function hasEnforcementFile(files: string[]): string[] {
  const enforcementFiles: string[] = [];
  for (const filePath of files) {
    for (const pattern of ENFORCEMENT_FILE_PATTERNS) {
      if (pattern.test(filePath)) {
        enforcementFiles.push(filePath);
        break;
      }
    }
  }
  return enforcementFiles;
}

async function main(): Promise<void> {
  let sessionId: string | undefined;
  try {
    const data = await parseHookInput();
    sessionId = data.session_id;
    const toolName = (data.tool_name as string) || "";

    // Only check Bash commands
    if (toolName !== "Bash") {
      await logHookExecution(HOOK_NAME, "approve", `not Bash: ${toolName}`, undefined, {
        sessionId,
      });
      const result = makeApproveResult(HOOK_NAME);
      console.log(JSON.stringify(result));
      return;
    }

    const toolInput = (data.tool_input as Record<string, unknown>) || {};
    const command = (toolInput.command as string) || "";

    // Check for skip environment variable
    if (isSkipEnvEnabled(process.env[SKIP_ENV_NAME])) {
      await logHookExecution(
        HOOK_NAME,
        "approve",
        `${SKIP_ENV_NAME} でスキップ（環境変数）`,
        undefined,
        { sessionId },
      );
      const result = makeApproveResult(HOOK_NAME);
      console.log(JSON.stringify(result));
      return;
    }

    const inlineValue = extractInlineSkipEnv(command, SKIP_ENV_NAME);
    if (isSkipEnvEnabled(inlineValue)) {
      await logHookExecution(
        HOOK_NAME,
        "approve",
        `${SKIP_ENV_NAME} でスキップ（インライン）`,
        undefined,
        { sessionId },
      );
      const result = makeApproveResult(HOOK_NAME);
      console.log(JSON.stringify(result));
      return;
    }

    // Check if this is a gh issue close command
    const issueNumber = extractIssueNumber(command);
    if (!issueNumber) {
      await logHookExecution(HOOK_NAME, "approve", "no issue number found", undefined, {
        sessionId,
      });
      const result = makeApproveResult(HOOK_NAME);
      console.log(JSON.stringify(result));
      return;
    }

    // Get issue content
    const content = getIssueContent(issueNumber);
    if (!content) {
      await logHookExecution(
        HOOK_NAME,
        "approve",
        `Issue #${issueNumber} の内容取得失敗`,
        undefined,
        { sessionId },
      );
      const result = makeApproveResult(HOOK_NAME);
      console.log(JSON.stringify(result));
      return;
    }

    const { title, body } = content;

    // Document update Issues are excluded (docs: or docs(...) prefix)
    if (/^docs[:\(]/i.test(title)) {
      await logHookExecution(
        HOOK_NAME,
        "approve",
        `Issue #${issueNumber} はドキュメント更新Issue（対象外）`,
        undefined,
        { sessionId },
      );
      const result = makeApproveResult(HOOK_NAME);
      console.log(JSON.stringify(result));
      return;
    }

    // Check for systematization keywords
    if (!hasSystematizationKeyword(title, body)) {
      await logHookExecution(
        HOOK_NAME,
        "approve",
        `Issue #${issueNumber} は仕組み化Issueではない`,
        undefined,
        { sessionId },
      );
      const result = makeApproveResult(HOOK_NAME);
      console.log(JSON.stringify(result));
      return;
    }

    // Get files changed in linked PRs
    const prFiles = getLinkedPrFiles(issueNumber);

    // Check for enforcement files
    const enforcementFiles = hasEnforcementFile(prFiles);

    if (enforcementFiles.length > 0) {
      let filesList = enforcementFiles.slice(0, 3).join(", ");
      if (enforcementFiles.length > 3) {
        filesList += ` 他${enforcementFiles.length - 3}件`;
      }
      await logHookExecution(
        HOOK_NAME,
        "approve",
        `Issue #${issueNumber} に強制機構ファイルあり: ${filesList}`,
        undefined,
        { sessionId },
      );
      const result = makeApproveResult(HOOK_NAME);
      console.log(JSON.stringify(result));
      return;
    }

    // No enforcement files - block
    const reasonLines = [
      `Issue #${issueNumber} は「仕組み化」を謳っていますが、`,
      "強制機構（フック/CI/ツール）の実装が確認できません。",
      "",
      "**仕組み化の定義**:",
      "  - ドキュメント追加だけでは不十分",
      "  - 違反を**ブロック**するフック/CI/ツールが必要",
      "",
      "**確認された変更ファイル**:",
    ];

    if (prFiles.length > 0) {
      for (const f of prFiles.slice(0, 5)) {
        reasonLines.push(`  - ${f}`);
      }
      if (prFiles.length > 5) {
        reasonLines.push(`  ... 他 ${prFiles.length - 5} 件`);
      }
    } else {
      reasonLines.push("  (関連PRが見つからないか、変更ファイルがありません)");
    }

    reasonLines.push(
      "",
      "**対応方法**:",
      "  1. フック/CI/スクリプトを実装してからクローズ",
      "  2. 強制機構が不要と判断した場合はコメントで理由を説明",
      "",
      "**スキップ方法（確認済みの場合）**:",
      `  SKIP_SYSTEMATIZATION_CHECK=1 gh issue close ${issueNumber}`,
    );

    const blockMessage = reasonLines.join("\n");

    await logHookExecution(
      HOOK_NAME,
      "block",
      `Issue #${issueNumber} に強制機構ファイルなし`,
      undefined,
      { sessionId },
    );

    const result = makeBlockResult(HOOK_NAME, blockMessage);
    console.log(JSON.stringify(result));
    process.exit(2);
  } catch (error) {
    // Don't block on errors - approve silently
    await logHookExecution(HOOK_NAME, "error", `フックエラー: ${formatError(error)}`, undefined, {
      sessionId,
    });
    const result = makeApproveResult(HOOK_NAME);
    console.log(JSON.stringify(result));
  }
}

if (import.meta.main) {
  main();
}
